test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java

```diff
@@ -0,0 +1,464 @@
+            StorageService.instance.stopNativeTransport();
+        });
+    };
+
+    public static InstanceAction disableBinary()
+    {
+        return disableBinary;
+    }
+
+    private static class DisseminateGossipState implements InstanceAction
+    {
+        final IInvokableInstance[] from;
+
+        public DisseminateGossipState(IInvokableInstance... from)
+        {
+            this.from = from;
+        }
+
+        public void apply(ICluster<IInvokableInstance> cluster, IInvokableInstance instance)
+        {
+            Map<InetSocketAddress, byte[]> m = new HashMap<>();
+            for (IInvokableInstance node : from)
+            {
+                byte[] epBytes = node.callsOnInstance(() -> {
+                    EndpointState epState = Gossiper.instance.getEndpointStateForEndpoint(FBUtilities.getBroadcastAddressAndPort());
+                    return toBytes(epState);
+                }).call();
+                m.put(node.broadcastAddress(), epBytes);
+            }
+
+            instance.appliesOnInstance((IIsolatedExecutor.SerializableFunction<Map<InetSocketAddress, byte[]>, Void>)
+                                       (map) -> {
+                                           Map<InetAddressAndPort, EndpointState> newState = new HashMap<>();
+                                           for (Map.Entry<InetSocketAddress, byte[]> e : map.entrySet())
+                                               newState.put(toCassandraInetAddressAndPort(e.getKey()), fromBytes(e.getValue()));
+
+                                           Gossiper.runInGossipStageBlocking(() -> {
+                                               Gossiper.instance.applyStateLocally(newState);
+                                           });
+                                           return null;
+                                       }).apply(m);
+        }
+    }
+
+    public static byte[] toBytes(EndpointState epState)
+    {
+        try (DataOutputBuffer out = new DataOutputBuffer(1024))
+        {
+            EndpointState.serializer.serialize(epState, out, MessagingService.current_version);
+            return out.toByteArray();
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static EndpointState fromBytes(byte[] bytes)
+    {
+        try (DataInputBuffer in = new DataInputBuffer(bytes))
+        {
+            return EndpointState.serializer.deserialize(in, MessagingService.current_version);
+        }
+        catch (Throwable t)
+        {
+            throw new RuntimeException(t);
+        }
+    }
+
+    private static class PullSchemaFrom implements InstanceAction
+    {
+        final IInvokableInstance pullFrom;
+
+        public PullSchemaFrom(IInvokableInstance pullFrom)
+        {
+            this.pullFrom = pullFrom;
+        }
+
+        public void apply(ICluster<IInvokableInstance> cluster, IInvokableInstance pullTo)
```

Comment:
The idea behind "actions" was in making such things work for multi-version clusters. We could make them named and each version could apply its corresponding version. But other than that - these actions _are_ in fact just functions.

Reply:
> The idea behind "actions" was in making such things work for multi-version clusters. 

But they don't currently as only one GossipHelper is available to the upgrade/mixed-version tests, so these classes/interfaces are in-fact single version only and unsafe in the multi-version cases.

> But other than that - these actions are in fact just functions.

Since these don't work in multi-version clusters why not just use functions?  While reading the tests I find the interfaces harder to understand as I need to figure out what they do, but relying on simple functions removes the ambiguity.

Reply:
> But they don't currently 

They don't. It was a forward-looking statement.

> why not just use functions?

You're suggesting to use static methods, not functions. I've explained motivation behind implementing an interface rather than functions above (uniformity, reusability, etc).

> interfaces harder to understand as I need to figure out what they do, but relying on simple functions removes the ambiguity.

This is simply not true. Both for a method call and an interface implementation you have to jump to definition to see what's really going on.